text/0000-calling-default-trait-methods.md

```diff
@@ -0,0 +1,277 @@
+
+Given a trait like:
+
+```rust
+pub trait Visit {
+    fn visit_block(&mut self, i: &Block) {
+        ...
+    }
+}
+```
+
+an impl can call the default implementation like so:
+
+```rust
+struct MyVisit;
+
+impl Visit for MyVisit {
+    fn visit_block(&mut self, i: &Block) {
+        self.super.visit_block(i);
+        // Alternatively:
+        <MyVisit as Visit>::super::visit_block(self, i);
+        ...
+    }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Method resolution in default implementations
+[method-resolution-in-default]: #method-resolution-in-default
+
+The default implementation will still call overridden implementations when calling trait methods, as it would normally. For recursive default implementations, the recursive call will resolve to the overriding implementation. For example, given:
+
+```rust
+trait Trait {
+    fn count_down(&self, n: usize) {
+        println!("Trait: {n}");
+        if n > 0 {
+            self.count_down(n - 1);
+        }
+    }
+}
+
+struct Struct;
+
+impl Trait for Struct {
+    fn count_down(&self, n: usize) {
+        println!("Struct: {n}");
+        self.super.count_down(n);
+    }  
+}
+```
+
+calling `Struct.count_down(2)` will print:
+
+```
+Struct: 2
+Trait: 2
+Struct: 1
+Trait: 1
+Struct: 0
+Trait: 0
+```
+
+## Permissive supercalls
+[permissive-supercalls]: #permissive-supercalls
+
+Supercalls are permitted anywhere, on the basis that the default implementation can only call other trait methods which the caller would be able to call anyway.
+
+## Interaction with specialization
+
+Supercalls always call the method which would have been called if the overriding implementation (and anything that overrides them) were not present. To specify other impls, universal function call syntax is extended to support naming specific impls.
+
+In general, impls are named by dropping `impl`, wrapping the signature in angle brackets, and specifying the values of the parameters. For example, `impl<T1, T2, ...> Trait for SomeType where T1: W1, T2: W2, ...` in a concrete context where `T1=C1`, `T2=C2`, ... can be named with `<<T1=C1, T2=C2, ...> Trait for SomeType where T1: W1, T2: W2, ...`. These match impls semantically rather than syntactically, i.e. `<<T: Display=Struct> Trait for T>` is equivalent to `<<T=Struct> Trait for T: Display>`.
+
+`<Struct as Trait>` names the most specific impl.
+
+Given:
+
+```
+trait Trait {
+  name(&self) -> &'static str {"Trait"}
+}
+
+impl<T: Display> Trait for T {
+  default name(&self) -> &'static str {"Display"}
+}
+
+impl<T: Display> Trait for Vec<T> {
+  default name(&self) -> &'static str {"Vec<Display>"}
+}
+
+impl Trait for Vec<String> {
+  name(&self) -> &'static str {"Vec<String>"}
+}
+```
+
+Then within a `Vec<String>` impl:
+
+- These evaluate to "Vec<String>":
```

Comment:
This doesn't render properly due to the angle brackets being interpreted as HTML. Put them in backticks instead.